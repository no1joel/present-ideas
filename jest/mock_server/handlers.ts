import { rest } from "msw";

export const handlers = [
  rest.get("http://localhost/api/their_list/:name", (req, res, context) => {
    return res(context.json({ presents: [] }));
  }),
  rest.get("http://localhost/api/my_list/:name", (req, res, context) => {
    return res(context.json({ presents: [] }));
  }),
];
