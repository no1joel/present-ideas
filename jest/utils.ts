import {
  render as testingLibraryRender,
  RenderOptions,
} from "@testing-library/vue";
import linkify from "vue-linkify";
import { createMemoryHistory, createRouter } from "vue-router";
import routes from "../present_ideas/present_ideas/static/router/routes";
import getStore from "../present_ideas/present_ideas/static/store";

const mockRouter = createRouter({ routes, history: createMemoryHistory() });

export const render = async (component, options = {}) => {
  const store = getStore();
  const defaultRenderOptions: RenderOptions = {
    global: {
      plugins: [store, mockRouter],
      directives: { linkified: linkify },
    },
  };
  const renderOptionsWithDefaults = { ...defaultRenderOptions, ...options };

  const result = testingLibraryRender(component, renderOptionsWithDefaults);

  return result;
};
