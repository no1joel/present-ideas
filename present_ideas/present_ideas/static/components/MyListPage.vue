<template>
  <Layout>
    <h1>This is gonna be your list...</h1>
  </Layout>
</template>

<script>
import Layout from "../layout/Layout.vue";

export default {
  components: { Layout },
};
</script>