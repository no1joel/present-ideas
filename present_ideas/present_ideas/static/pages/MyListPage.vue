<template>
  <div class="container-fluid">
    <div class="row">
      <div class="col">
        <LoadingIndicator v-if="loading" />
        <div class="card-columns">
          <PresentIdeaFormCard
            v-bind:user="$route.params.user"
            v-bind:added-by="$route.params.user"
            v-on:added="fetchMyList"
          />
          <PresentIdeaCard
            v-for="present in presents"
            v-bind="present"
            v-bind:key="
              present.index + present.Thing + present.Price + present.Notes
            "
            v-bind:loading="loading"
            v-on:delete-clicked="deleteClicked"
            v-on:update-name="updateName"
            v-on:update-price="updatePrice"
            v-on:update-notes="updateNotes"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import LoadingIndicator from "../components/LoadingIndicator.vue";
import PresentIdeaCard from "../components/PresentIdeaCard.vue";
import PresentIdeaFormCard from "../components/PresentIdeaFormCard.vue";
import Layout from "../layout/Layout.vue";

export default {
  components: {
    Layout,
    PresentIdeaCard,
    PresentIdeaFormCard,
    LoadingIndicator,
  },
  data() {
    return { presents: [], loading: false };
  },
  mounted() {
    this.fetchMyList();
  },
  computed: {
    user() {
      return this.$route.params.user;
    },
    baseUrl() {
      return window.location.origin || "http://localhost";
    },
  },
  methods: {
    async fetchMyList() {
      this.loading = true;

      const url = `${this.baseUrl}/api/my_list/${this.user}`;
      const response = await fetch(url);
      const data = await response.json();
      const presents = data.presents;
      this.presents = presents.reverse();

      this.loading = false;
    },
    async deleteClicked({ index }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/delete_idea/`;
      const data = {
        index: index,
        user: this.user,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
    async updateName({ index, name }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/update_idea_name/`;
      const data = {
        index: index,
        user: this.user,
        name: name,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
    async updatePrice({ index, price }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/update_idea_price/`;
      const data = {
        index: index,
        user: this.user,
        price,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
    async updateNotes({ index, notes }) {
      this.loading = true;
      const url = `${this.baseUrl}/api/update_idea_notes/`;
      const data = {
        index: index,
        user: this.user,
        notes,
      };
      await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });
      this.fetchMyList();
    },
  },
};
</script>
