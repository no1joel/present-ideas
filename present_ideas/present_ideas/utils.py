from typing import Iterable, List

from sheets.spreadsheet import (
    add_row,
    delete_row,
    get_rows,
    get_worksheets,
    set_cell_value,
)


def check_name(name: str) -> bool:
    """Check if the provided name is a sheet."""

    worksheet_names = get_all_names()
    return name.lower() in worksheet_names


def get_all_names() -> Iterable[str]:
    """Return people's names."""

    return (
        worksheet.title.lower()
        for worksheet in get_worksheets()
        if "template" not in worksheet.title.lower()
    )


def get_present_ideas(their_name: str) -> List[dict]:
    # TODO: Embed urls
    _header, *rows = get_rows(their_name)
    presents = [{**row, "index": index} for index, row in enumerate(rows)]
    return presents


def get_row_index(thing_index: int) -> int:
    # TODO: Why is it plus 3?!
    return thing_index + 3
