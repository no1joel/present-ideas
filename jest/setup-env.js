// test/setup-env.js
// add this to your setupFilesAfterEnv config in jest so it's imported for every test file
import fetch, { Headers, Request, Response } from "node-fetch";
import { TextDecoder as UtilTextDecoder } from "util";
import { server } from "../mock_server";

if (!globalThis.fetch) {
  // @ts-ignore
  globalThis.fetch = fetch;
  globalThis.Headers = Headers;
  // @ts-ignore
  globalThis.Request = Request;
  // @ts-ignore
  globalThis.Response = Response;
}

if (!globalThis.TextDecoder) {
  globalThis.TextDecoder = UtilTextDecoder;
}

beforeAll(() => server.listen());
// if you need to add a handler after calling setupServer for some specific test
// this will remove that handler for the rest of them
// (which is important for test isolation):
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
