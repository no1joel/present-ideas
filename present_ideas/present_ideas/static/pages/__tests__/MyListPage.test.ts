import { screen, waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";
import { server } from "../../../../../jest/mock_server/server";
import { render } from "../../../../../jest/utils";
import MyListPage from "../MyListPage.vue";

describe("MyListPage", () => {
  const presents: { Thing: string; Notes: string; Price: number }[] = [];
  beforeEach(() => {
    server.use(
      rest.get("http://localhost/api/my_list/:name", (req, res, ctx) => {
        return res(
          ctx.json({
            presents: presents.map((present, index) => ({ ...present, index })),
          })
        );
      }),
      rest.post("http://localhost/api/delete_idea/", async (req, res, ctx) => {
        const { index } = await req.json();
        presents.splice(index, 1);
        return res(ctx.json({}));
      })
    );
  });
  afterEach(() => {
    presents.splice(0);
  });
  it("matches snapshot", async () => {
    const screen = await render(MyListPage);
    expect(screen.container).toMatchSnapshot();
  });
  it("removes when you click delete", async () => {
    // TODO: Need to set up route to be current user's name and assert that
    // the name is there
    const user = userEvent.setup();
    presents.push({
      Thing: "A present",
      Notes: "asdf",
      Price: 100,
    });
    await render(MyListPage);

    const deleteButton = await waitFor(() => screen.getAllByText("Delete")[0]);
    await user.click(deleteButton);
    const yesButton = await waitFor(() => screen.getAllByText("Yep!")[0]);
    await user.click(yesButton);
    await waitFor(() =>
      expect(screen.queryAllByText("Delete")).toHaveLength(0)
    );
  });
  it("removes the correct one when you click delete", async () => {
    const user = userEvent.setup();
    presents.push({
      Thing: "first present",
      Notes: "",
      Price: 1,
    });
    presents.push({
      Thing: "second present",
      Notes: "",
      Price: 2,
    });
    presents.push({
      Thing: "third present",
      Notes: "",
      Price: 3,
    });
    await render(MyListPage);

    await waitFor(() =>
      expect(screen.getAllByRole("button", { name: "Delete" })).toHaveLength(3)
    );
    const deleteButtons = screen.getAllByRole("button", { name: "Delete" });
    await user.click(deleteButtons[0]);
    const yesButtons = await screen.findAllByText("Yep!");
    await user.click(yesButtons[0]);
    await waitFor(() =>
      expect(screen.queryAllByText("Delete")).toHaveLength(2)
    );
    expect(screen.queryByText("third present")).toBeNull();
    screen.getByText("second present");
    screen.getByText("first present");
  });
});
