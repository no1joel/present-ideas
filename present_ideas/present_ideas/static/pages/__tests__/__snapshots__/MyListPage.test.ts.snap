// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyListPage > matches snapshot 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="d-flex justify-content-center align-items-center"
          style="background: rgba(255, 255, 255, 0.4); position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px;"
        >
          <div
            class="spinner-border"
            role="status"
          >
            <span
              class="sr-only"
            >
              Loading...
            </span>
          </div>
        </div>
        <div
          class="card-columns"
        >
          <div
            class="card"
          >
            <!--v-if-->
            <div
              class="card-header"
            >
              <h5
                class="card-title m-0"
              >
                Add New!
              </h5>
              <!--v-if-->
            </div>
            <div
              class="card-body"
            >
              <form
                class="form-horizontal"
              >
                <div
                  class="form-group"
                >
                  <label
                    for="thing-input"
                  >
                    Thing*
                  </label>
                  <input
                    class="form-control"
                    id="thing-input"
                    name="thing"
                    placeholder="Something I want..."
                    type="text"
                  />
                </div>
                <div
                  class="form-group"
                >
                  <label
                    for="price-input"
                  >
                    Price
                  </label>
                  <input
                    class="form-control"
                    id="price-input"
                    name="price"
                    placeholder="Enter a price or a range, e.g. ¬£10-¬£20"
                    type="text"
                  />
                </div>
                <div
                  class="form-group"
                >
                  <label
                    for="notes-input"
                  >
                    Notes
                  </label>
                  <textarea
                    class="form-control"
                    id="notes-input"
                    name="notes"
                    placeholder="Enter a link to a specific thing, a description of what you want, why you want it, anything!"
                  />
                </div>
                <button
                  class="btn btn-success btn-block"
                  disabled=""
                  type="button"
                >
                   Add 
                </button>
              </form>
            </div>
          </div>
          
          
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MyListPage > removes when you click delete 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <!--v-if-->
        <div
          class="card-columns"
        >
          <div
            class="card"
          >
            <!--v-if-->
            <div
              class="card-header"
            >
              <h5
                class="card-title m-0"
              >
                Add New!
              </h5>
              <!--v-if-->
            </div>
            <div
              class="card-body"
            >
              <form
                class="form-horizontal"
              >
                <div
                  class="form-group"
                >
                  <label
                    for="thing-input"
                  >
                    Thing*
                  </label>
                  <input
                    class="form-control"
                    id="thing-input"
                    name="thing"
                    placeholder="Something I want..."
                    type="text"
                  />
                </div>
                <div
                  class="form-group"
                >
                  <label
                    for="price-input"
                  >
                    Price
                  </label>
                  <input
                    class="form-control"
                    id="price-input"
                    name="price"
                    placeholder="Enter a price or a range, e.g. ¬£10-¬£20"
                    type="text"
                  />
                </div>
                <div
                  class="form-group"
                >
                  <label
                    for="notes-input"
                  >
                    Notes
                  </label>
                  <textarea
                    class="form-control"
                    id="notes-input"
                    name="notes"
                    placeholder="Enter a link to a specific thing, a description of what you want, why you want it, anything!"
                  />
                </div>
                <button
                  class="btn btn-success btn-block"
                  disabled=""
                  type="button"
                >
                   Add 
                </button>
              </form>
            </div>
          </div>
          
          <div
            class="card"
            id="1"
            name="A present"
            notes="asdf"
            price="100"
          >
            <!--v-if-->
            <div
              class="card-header"
            >
              <h5
                class="card-title m-0"
              />
              <small>
                <p
                  class="card-text text-right text-muted"
                >
                  <span>
                    Price: 
                    <span>
                      ü§∑‚Äç‚ôÄÔ∏è
                    </span>
                  </span>
                  <!--v-if-->
                </p>
                <!--v-if-->
              </small>
              <p
                class="card-text text-right mt-1"
              >
                <div
                  class="inline-block"
                  data-v-5784ed69=""
                  style="border: 12px solid transparent; margin: -12px;"
                >
                  <div
                    data-v-5784ed69=""
                  >
                    
                    <button
                      class="btn btn-sm btn-outline-danger"
                      data-v-5784ed69-s=""
                      slot="reference"
                      type="button"
                    >
                       Delete 
                    </button>
                    
                  </div>
                  <transition-stub
                    appear="false"
                    css="true"
                    data-v-5784ed69=""
                    name="fade"
                    persisted="false"
                  >
                    <div
                      class="popper"
                      data-v-5784ed69=""
                      style="display: none;"
                    >
                      
                      <div
                        class="popper popover"
                        data-v-5784ed69-s=""
                      >
                        <div
                          class="popover-header"
                          data-v-5784ed69-s=""
                        >
                          Really delete?
                        </div>
                        <div
                          class="popover-body"
                          data-v-5784ed69-s=""
                        >
                          <button
                            class="btn btn-block btn-danger"
                            data-v-5784ed69-s=""
                            type="button"
                          >
                             Yep! 
                          </button>
                        </div>
                      </div>
                      
                      <!---->
                    </div>
                  </transition-stub>
                </div>
              </p>
            </div>
            <div
              class="card-body"
            >
              
              <p
                class="text-muted font-italic text-center"
              >
                No notes!
              </p>
              
              <!--v-if-->
            </div>
            <!--v-if-->
          </div>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
